Graphics/Slicer/Math/PGA.hs

```diff
@@ -74,50 +67,126 @@ plinesIntersectIn pl1 pl2
+-- | Find out where two lines intersect, returning a projective point.
+intersectionOf :: PLine2 -> PLine2 -> PPoint2
+intersectionOf pl1 pl2 = canonicalizePPoint2 $ meet2PLine2 pl1 pl2
+
+-- | Find... something?
+--   FIXME: this is a total hack. returns positive for left, negaative for right, and +45 degrees returns the same as -45 degrees.
 dualAngle :: GVec -> GVec -> ℝ
 dualAngle ln1 ln2 = valOf 0 $ getVals [GEZero 1, GEZero 1, GEPlus 1, GEPlus 2] $ (\(GVec a) -> a) $ unlikeVecPair ln1 ln2
 
--- | find the angle of intersection of two PLines.
+-- | Find the angle of intersection of two PLines.
 angleBetween :: PLine2 -> PLine2 -> ℝ
 angleBetween pl1 pl2 =  scalarPart $ pv1 ⎣ pv2
   where
     (PLine2 pv1) = forcePLine2Basis (normalizePLine2 pl1)
     (PLine2 pv2) = forcePLine2Basis (normalizePLine2 pl2)
 
+-- | Translate a line a given amound along it's perpendicular bisector.
+translatePerp :: PLine2 -> ℝ -> PLine2
+translatePerp pl1 d = PLine2 $ addVecPair m $ rawPLine pl1
+  where
+    m = GVec [GVal (d*normOfPLine2 pl1) [GEZero 1]]
+    rawPLine (PLine2 a) = a
+
+-- | find the distance between a point and a line.
+distancePPointToPLine :: PPoint2 -> PLine2 -> ℝ
+distancePPointToPLine point line = normOfPLine2 $ join2PPoint2 point linePoint
+--error $ "result: " <> show (normOfPLine2 $ newLine) <> "\nLine: " <> show lvec <> "\nPoint: " <> show pvec <> "\nPerpLine: " <> show perpLine <> "\nLinePoint: " <> show linePoint <> "\nnewLine: " <> show newLine <> "\n"
+  where
+    (PLine2 lvec)  = normalizePLine2 line
+    (PPoint2 pvec) = canonicalizePPoint2 point
+    perpLine       = PLine2 $ lvec ⨅ pvec
+    linePoint      = meet2PLine2 (PLine2 lvec) perpLine
+    newLine        = join2PPoint2 linePoint (PPoint2 pvec)
+
+----------------------------------------------------------
+-------------- Euclidian Mixed Interface -----------------
+----------------------------------------------------------
+
+-- | The Linear result of line segment intersection in 2 dimensions.
+data Intersection =
+    NoIntersection
+  | HitStartPointL2 Point2
+  | HitEndPointL2 Point2
+  deriving (Show)
+
+-- | an alias, for cases where either input is acceptable.
+type SegOrPLine2 = Either LineSeg PLine2
+
+-- Check if/where lines/line segments intersect.
+-- entry point usable for all intersection needs, but it's faster to use the above two functions when you KNOW there MUST be an intersection.
+intersectsWith :: SegOrPLine2 -> SegOrPLine2 -> Either Intersection PIntersection
+intersectsWith (Left l1)   (Left l2)   =         lineIntersection    l1  l2
+intersectsWith (Right pl1) (Right pl2) = Right $ plinesIntersectIn   pl1 pl2
+intersectsWith (Left l1)   (Right pl1) =         lineIntersectsPLine l1  pl1
+intersectsWith (Right pl1) (Left l1)   =         lineIntersectsPLine l1  pl1
+
+-- Check if/where lines/line segments intersect.
+lineIntersectsPLine :: LineSeg -> PLine2 -> Either Intersection PIntersection
```

Comment:
It seems to me like you want an actual `data` for this `Either` type, seeing as it keeps popping up everywhere.

Reply:
And this one is trying to be an (optional) union type.